hedera-node/src/main/java/com/hedera/services/queries/answering/AnswerFunctions.java

```diff
@@ -21,54 +21,110 @@
+				? recordsFrom(targetAccount, numAvailable)
+				: mostRecentFrom(targetAccount, maxQueryable, numAvailable);
 	}
 
-	public Optional<TransactionRecord> txnRecord(
-			final RecordCache recordCache,
-			final StateView view,
-			final TransactionGetRecordQuery query
-	) {
-		final var txnId = query.getTransactionID();
+	/**
+	 * Returns the record of the requested transaction from the given {@link RecordCache}, if available.
+	 *
+	 * @param recordCache the cache to get the record from
+	 * @param op the query with the target transaction id
+	 * @return the transaction record if available
+	 */
+	public Optional<TransactionRecord> txnRecord(final RecordCache recordCache, final TransactionGetRecordQuery op) {
+		final var txnId = op.getTransactionID();
 		final var expirableTxnRecord = recordCache.getPriorityRecord(txnId);
-		if (expirableTxnRecord != null) {
-			return Optional.of(expirableTxnRecord.asGrpc());
-		} else {
-			try {
-				final var id = txnId.getAccountID();
-				final var account = view.accounts().get(EntityNum.fromAccountId(id));
-				final var searchableId = TxnId.fromGrpc(txnId);
-				return account.recordList()
-						.stream()
-						.filter(r -> r.getTxnId().equals(searchableId))
-						.findAny()
-						.map(ExpirableTxnRecord::asGrpc);
-			} catch (final Exception ignore) {
-				return Optional.empty();
+		return Optional.ofNullable(expirableTxnRecord).map(ExpirableTxnRecord::asGrpc);
+	}
+
+	/* --- Internal helpers --- */
+	/**
+	 * Returns up to {@code n} payer records from an account in gRPC form.
+	 *
+	 * @param account the account of interest
+	 * @param n the expected number of records in the account
+	 * @return the available records
+	 */
+	private List<TransactionRecord> recordsFrom(final MerkleAccount account, final int n) {
+		return mostRecentFrom(account, n, n);
+	}
+
+	/**
+	 * Returns up to the last {@code m}-of-{@code n} payer records from an account in gRPC form.
+	 *
+	 * Since records are added FIFO to the payer account, the last records are the most recent;
+	 * and presumably the most interesting.
+	 *
+	 * If the given {@link MerkleAccount} is from the working state (as in release 0.22), then
+	 * this method acts on a best-effort basis, and returns only the relevant records it could
+	 * iterate over before hitting a {@link ConcurrentModificationException} or {@link NoSuchElementException}.
+	 *
+	 * @param account the account of interest
+	 * @param m the maximum number of records to return
+	 * @param n the expected number of records in the account
+	 * @return the available records
+	 */
+	private List<TransactionRecord> mostRecentFrom(final MerkleAccount account, final int m, final int n) {
+		final List<TransactionRecord> ans = new ArrayList<>();
+		final Iterator<ExpirableTxnRecord> iter = account.recordIterator();
+		try {
+			for (int i = 0, cutoff = n - m; i < n; i++) {
+				final var nextRecord = iter.next();
+				if (i >= cutoff) {
+					ans.add(nextRecord.asGrpc());
+				}
 			}
```

Comment:
Ouch, this looks inefficient. It looks like the platform could easily add a reverse iterator, which would keep you from having to iterate over the entire body of records in this case.

Reply:
I filed an issue with Swirlds. We can come back and look at making this more efficient after reverse iteration is supported.